src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestArrayUpdate.ll

```diff
@@ -7,9 +7,11 @@ entry:
   %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %1, i64 %a)
   %3 = bitcast i8* %2 to i64*
   store i64 %b, i64* %3
+  %4 = load %Array*, %Array** %x
```

Comment:
```
define %Array* @Microsoft__Quantum__Testing__QIR__TestArrayUpdate__body(%Array* %y, i64 %a, i64 %b) {
entry:
  %x = alloca %Array*
  store %Array* %y, %Array** %x
  %0 = load %Array*, %Array** %x
  %1 = call %Array* @__quantum__rt__array_copy(%Array* %0) ; could use %y instead %0
  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %1, i64 %a)
  %3 = bitcast i8* %2 to i64*
  store i64 %b, i64* %3
  %4 = load %Array*, %Array** %x ; dead load, because no reads between it and the next load into %4
  store %Array* %1, %Array** %x
  call void @__quantum__rt__array_reference(%Array* %1)
  %4 = load %Array*, %Array** %x ; %4 == %1
  %5 = load %Array*, %Array** %x ; %5 == %1
  call void @__quantum__rt__array_unreference(%Array* %1) ; this pair of addref/release serves no purpose
  ret %Array* %4
  ; below is unreachable because there are no branches in this function, 
  ; but if it was, it would return a dangling pointer, because it would release the copied 
  ; array via %4 and then return it via %5
  call void @__quantum__rt__array_unreference(%Array* %4)
  ret %Array* %5
}
```
I would expect the IR that implements the same functionality to look more like this:
```
define %Array* @Microsoft__Quantum__Testing__QIR__TestArrayUpdate__body(%Array* %y, i64 %a, i64 %b) {
entry:
  %1 = call %Array* @__quantum__rt__array_copy(%Array* %y)
  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %1, i64 %a)
  %3 = bitcast i8* %2 to i64*
  store i64 %b, i64* %3
  ret %Array* %1
}
```

Reply:
Indeed, the LLVM given is actually illegal because it violates SSA (%4 gets assigned to twice). What @irinayat-MS suggests looks correct to me.

Reply:
Hu, I am not sure how it is even possible to generate that code - the variable name/numbering is automatically generated by the builder, not by me.